### PR TITLE
Fix APYs and add support for single sided LPs

### PR DESCRIPTION
**Pull request by clabby**
Fixes APYs on FarmCards using ChainLink oracles.

preview

### CODE REVIEW

File: src/panda/utils.ts [hunk 1]
```diff
@@ -124,6 +131,134 @@ export const getLockedEarned = async (
...
+  bigNumber.div(new BigNumber(10).pow(decimals))
+
+interface PriceOracle {
+  token: string,
+  address: string,
+  contract: Contract,
+}
+
+const getOraclePrice = async (tokenSymbol: string, priceOracles: Array<PriceOracle>) => {
+  const oracle = _.find(priceOracles, { token: tokenSymbol })
```
Comment (jondwillis) [inline src/panda/utils.ts:144]: I think lodash is a big lib to import. you can do similar with native: https://riptutorial.com/typescript/example/29544/finding-object-in-array

File: src/panda/utils.ts [hunk 2]
```diff
@@ -124,6 +131,134 @@ export const getLockedEarned = async (
...
+) => {
+  const [resultA, resultB, resultC, resultD, resultE] = await Promise.all([
+    getTotalLPUSDValue(0, masterChefContract, pnda, true),
+    getTotalLPUSDValue(1, masterChefContract, pnda, true),
+    getTotalLPUSDValue(2, masterChefContract, pnda, true),
+    getTotalLPUSDValue(3, masterChefContract, pnda, true),
+    getTotalLPUSDValue(4, masterChefContract, pnda, true),
+  ])
+
+  return resultA.lockedUsd
```
Comment (jondwillis) [inline src/panda/utils.ts:166]: you could put these in an array, `reduce +`, then divide by the original array `count`

File: src/panda/utils.ts [hunk 3]
```diff
@@ -124,6 +131,134 @@ export const getLockedEarned = async (
...
+  reward: BigNumber
+}> => {
+  const { web3 } = pnda
+  const supportedPools = getFarms(pnda)
+  const pool = _.find(supportedPools, { pid })
+  const { lpContract } = pool
+  const priceOracles = oracles(web3)
+
+  // Special case: Single asset LP
+  if (pool.pid === 5 || pool.pid === 6) {
```
Comment (jondwillis) [inline src/panda/utils.ts:191]: prefer detecting this instead of hardcoding, you could also extract this to an array `const`

**Review by jondwillis — CHANGES_REQUESTED** [review]
looks really good, left a few comments

**Comment by clabby** [comment]
> looks really good, left a few comments

Great suggestions. Just pushed a commit that removes the lodash dependency, reduces the Panda Price calls rather than manually adding / averaging them, and detects single asset LPs by checking if the associated LP contract address is equal to the token address rather than a hard-coded rule.

**Review by jondwillis — APPROVED** [review]

File: yarn.lock [hunk 4]
```diff
@@ -2029,6 +2029,11 @@
   resolved "https://registry.yarnpkg.com/@types/json5/-/json5-0.0.29.tgz#ee28707ae94e11d2b827bcbe5270bcea7f3e71ee"
   integrity sha1-7ihweulOEdK4J7y+UnC86n8+ce4=
 
+"@types/lodash@^4.14.169":
```
Comment (jondwillis) [inline yarn.lock:2032]: oh remove this too

**Review by jondwillis — CHANGES_REQUESTED** [review]

**Comment by jondwillis** [comment]
Also, check Pancakeswap LP's.. the APYs do not seem correct.

**Comment by clabby** [comment]
> Also, check Pancakeswap LP's.. the APYs do not seem correct.

Forgot to add % staked into the consideration, oops. Should be fixed.

a
a

File: src/hooks/useAllStakedValue.ts [hunk 5]
```diff
@@ -8,18 +8,23 @@ import {
...
   getFarms,
-  getTotalLPWbnbValue,
+  getTotalLPUSDValue,
 } from '../panda/utils'
 import usePanda from './usePanda'
 import useBlock from './useBlock'
-import { getContract } from '../utils/erc20'
 
-export interface StakedValue {
+/*export interface StakedValue {
```
Comment (jondwillis) [inline src/hooks/useAllStakedValue.ts:16]: remove commented out code

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,4 @@ node_modules
 build/
 *.log
 .env
+package-lock.json
```
Comment (jondwillis) [inline .gitignore:7]: remove this

**Review by jondwillis — APPROVED** [review]

File: src/hooks/useAllStakedValue.ts (hunk 5 above)
Comment (clabby) [reply src/hooks/useAllStakedValue.ts]: https://github.com/baofinance/ui-panda-farms/pull/26/commits/5e82bdeaee4515d9e149f32d8e7b96b43e6a6259#diff-cf29e8651feb9fbb11338301534e63721dea5a932b8585cc0e8d30b6a9eaf9df

File: .gitignore (hunk 6 above)
Comment (clabby) [reply .gitignore]: https://github.com/baofinance/ui-panda-farms/pull/26/commits/5e82bdeaee4515d9e149f32d8e7b96b43e6a6259#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947

**Pull request merged**